### ISSUE TITLE
update a column present in multiple tables in dotnet EntityFrameworkCore

### ISSUE THREAD

**Issue by nagasudhirpulla** [nagasudhirpulla/GridElementsService#2]
I am using EntityFrameworkCore version 8

There are multiple tables (declared as DBSet<TableClass>) present in the DbContext having the column "OwnersCache"

I would like to update the values in OwnersCache column of all the database tables without specifying the table names explicitly
How to implement this with the DbContext...

To be more precise, I have a variable `dbContext`. The substring `Name1` in the OwnersCache column of all the database tables should be replaced with `NewName1`

question raised in https://stackoverflow.com/q/79219285/2746323 and https://www.reddit.com/r/dotnet/comments/1gyhyra/update_a_column_present_in_multiple_tables_in/?rdt=36259

**Comment by nagasudhirpulla** [comment]
Get all entity types containing the column

```csharp
 var entityTypes = dbContext.Model.GetEntityTypes()
    .Where(e => e.GetProperties().Any(p => p.Name == "OwnersCache"))
    .ToList();

```

**Comment by nagasudhirpulla** [comment]
code from claude
```csharp
using Microsoft.EntityFrameworkCore;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore.Metadata;
using System.Reflection;

public static class DbContextExtensions
{
    public static async Task UpdateOwnersCacheAcrossAllTables(this DbContext dbContext, string oldValue, string newValue)
    {
        // Get all entity types from the model that have OwnersCache property of type string
        var entityTypes = dbContext.Model.GetEntityTypes()
            .Where(e => e.GetProperties()
                .Any(p => p.Name == "OwnersCache" && 
                         p.ClrType == typeof(string)))
            .ToList();

        if (!entityTypes.Any())
        {
            throw new InvalidOperationException("No entities found with string property named 'OwnersCache'");
        }

        foreach (var entityType in entityTypes)
        {
            // Get the CLR type for the entity
            var clrType = entityType.ClrType;

            // Verify the property type at CLR level as well
            var ownersCacheProperty = clrType.GetProperty("OwnersCache");
            if (ownersCacheProperty == null || ownersCacheProperty.PropertyType != typeof(string))
            {
                continue; // Skip if property is not found or not string type
            }

            // Get the DbSet dynamically using reflection
            var dbSetProperty = dbContext.GetType().GetProperties()
                .FirstOrDefault(p => p.PropertyType.IsGenericType &&
                                   p.PropertyType.GetGenericTypeDefinition() == typeof(DbSet<>) &&
                                   p.PropertyType.GetGenericArguments()[0] == clrType);

            if (dbSetProperty != null)
            {
                try
                {
                    // Get the DbSet instance
                    var dbSet = dbSetProperty.GetValue(dbContext);
                    
                    // Create generic method to handle the update
                    var updateMethod = typeof(DbContextExtensions)
                        .GetMethod(nameof(UpdateOwnersCacheForType), BindingFlags.NonPublic | BindingFlags.Static)
                        .MakeGenericMethod(clrType);

                    // Invoke the generic update method
                    await (Task)updateMethod.Invoke(null, new[] { dbSet, oldValue, newValue });
                }
                catch (Exception ex)
                {
                    throw new Exception($"Failed to update OwnersCache for entity {entityType.Name}", ex);
                }
            }
        }

        // Save all changes
        await dbContext.SaveChangesAsync();
    }

    private static async Task UpdateOwnersCacheForType<TEntity>(object dbSet, string oldValue, string newValue) 
        where TEntity : class
    {
        var set = (DbSet<TEntity>)dbSet;
        
        // Double check the property type at runtime
        var property = typeof(TEntity).GetProperty("OwnersCache");
        if (property == null || property.PropertyType != typeof(string))
        {
            throw new InvalidOperationException($"Property OwnersCache of type {typeof(TEntity).Name} is not a string");
        }

        var entities = await set
            .Where(e => EF.Property<string>(e, "OwnersCache").Contains(oldValue))
            .ToListAsync();

        foreach (var entity in entities)
        {
            var currentValue = (string)property.GetValue(entity);
            if (currentValue != null)
            {
                property.SetValue(entity, currentValue.Replace(oldValue, newValue));
            }
        }
    }
}
```

**Comment by nagasudhirpulla** [comment]
test generated by claude
```csharp
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using System.Linq;

// Entity classes
public class Customer
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string OwnersCache { get; set; }
}

public class Product
{
    public int Id { get; set; }
    public string Name { get; set; }
    public string OwnersCache { get; set; }
}

public class Order
{
    public int Id { get; set; }
    public DateTime OrderDate { get; set; }
    public string OwnersCache { get; set; }
}

// Invalid entity to test type checking
public class InvalidEntity
{
    public int Id { get; set; }
    public int OwnersCache { get; set; }  // Integer instead of string
}

// DbContext
public class TestDbContext : DbContext
{
    public DbSet<Customer> Customers { get; set; }
    public DbSet<Product> Products { get; set; }
    public DbSet<Order> Orders { get; set; }
    public DbSet<InvalidEntity> InvalidEntities { get; set; }

    public TestDbContext(DbContextOptions<TestDbContext> options) : base(options)
    {
    }

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        base.OnModelCreating(modelBuilder);
        
        // Seed data
        modelBuilder.Entity<Customer>().HasData(
            new Customer { Id = 1, Name = "Customer 1", OwnersCache = "Name1,Admin" },
            new Customer { Id = 2, Name = "Customer 2", OwnersCache = "Name1,User" }
        );

        modelBuilder.Entity<Product>().HasData(
            new Product { Id = 1, Name = "Product 1", OwnersCache = "Name1,Manager" },
            new Product { Id = 2, Name = "Product 2", OwnersCache = "Name2,Admin" }
        );

        modelBuilder.Entity<Order>().HasData(
            new Order { Id = 1, OrderDate = DateTime.Now, OwnersCache = "Name1,Sales" },
            new Order { Id = 2, OrderDate = DateTime.Now, OwnersCache = null }  // Test null handling
        );

        modelBuilder.Entity<InvalidEntity>().HasData(
            new InvalidEntity { Id = 1, OwnersCache = 123 }
        );
    }
}

// Test class
public class OwnersCacheUpdateTests
{
    private readonly TestDbContext _dbContext;

    public OwnersCacheUpdateTests()
    {
        var options = new DbContextOptionsBuilder<TestDbContext>()
            .UseInMemoryDatabase(databaseName: "TestDb_" + Guid.NewGuid().ToString())
            .Options;

        _dbContext = new TestDbContext(options);
        _dbContext.Database.EnsureCreated();
    }

    public async Task RunTests()
    {
        try
        {
            // Print initial state
            Console.WriteLine("Initial State:");
            await PrintCurrentState();

            // Perform update
            Console.WriteLine("\nUpdating OwnersCache values...");
            await _dbContext.UpdateOwnersCacheAcrossAllTables("Name1", "NewName1");

            // Print final state
            Console.WriteLine("\nFinal State:");
            await PrintCurrentState();
        }
        catch (Exception ex)
        {
            Console.WriteLine($"\nError during test: {ex.Message}");
        }
    }

    private async Task PrintCurrentState()
    {
        var customers = await _dbContext.Customers.ToListAsync();
        var products = await _dbContext.Products.ToListAsync();
        var orders = await _dbContext.Orders.ToListAsync();

        Console.WriteLine("Customers:");
        foreach (var customer in customers)
        {
            Console.WriteLine($"ID: {customer.Id}, Name: {customer.Name}, OwnersCache: {customer.OwnersCache}");
        }

        Console.WriteLine("\nProducts:");
        foreach (var product in products)
        {
            Console.WriteLine($"ID: {product.Id}, Name: {product.Name}, OwnersCache: {product.OwnersCache}");
        }

        Console.WriteLine("\nOrders:");
        foreach (var order in orders)
        {
            Console.WriteLine($"ID: {order.Id}, OwnersCache: {order.OwnersCache}");
        }
    }
}

// Program to run the test
public class Program
{
    public static async Task Main()
    {
        var tests = new OwnersCacheUpdateTests();
        await tests.RunTests();
    }
}
```

**Comment by nagasudhirpulla** [comment]
fixed in 6021ffaeb22783df808c74014509c0b7e8204bec
